### PR TITLE
Queue identification driver

### PR DESCRIPTION
**Pull request by luceos**

### CODE REVIEW

File: src/Identification/Queue/Events/Processing.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        $environment->setTenant($tenant ?? $this->tenant() ?? $this->resolve());
+
+        return $this;
+    }
+
+    protected function resolver(): ResolvesTenants
+    {
+        return resolve(ResolvesTenants::class);
+    }
```
Comment (ArlonAntonius) [inline src/Identification/Queue/Events/Processing.php]: Like I said before, I would like to get rid of all these functions. I don't think that's the approach we should take.

File: src/Identification/Queue/Providers/IdentificationProvider.php [hunk 2]
```diff
@@ -16,54 +16,28 @@
...
+            Queue::createPayloadUsing($this->app->make(Middleware\SaveTenantOnQueuePayload::class));
 
-                    $environment->setTenant($tenant);
-                }
-            });
+            // Resolve any tenant related meta data on job and allow resolving of tenant.
+            $queue->before($this->app->make(Middleware\ReadTenantFromQueuePayload::class));
 
             return $queue;
         });
```
Comment (ArlonAntonius) [inline src/Identification/Queue/Providers/IdentificationProvider.php]: Why is it first using the facade and after that using the provided variable?

File: tests/unit/Identification/Queue/Mocks/TenantIdentifiableInQueue.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+     * Specify whether the tenant model is matching the queue job.
+     *
+     * @param Processing $event
+     *
+     * @return Tenant
+     */
+    public function tenantIdentificationByQueue(Processing $event = null): ?Tenant
+    {
+        return $this->newQuery()->first();
+    }
```
Comment (ArlonAntonius) [inline tests/unit/Identification/Queue/Mocks/TenantIdentifiableInQueue.php]: Let's put some actual logic here for testing. Simply returning first is not a real use case.
Check the HTTP Identification Driver Tests, you didn't simply return first there either.

File: src/Tenancy/Identification/TenantResolver.php [hunk 4]
```diff
@@ -153,7 +154,13 @@ protected function resolveFromDrivers(TenantModelCollection $models): ?Tenant
...
                     /** @var ReflectionMethod $method */
                     foreach ($this->retrieveDriverMethods($driver) as $method) {
-                        if ($tenant = app()->call("$item@{$method->getName()}")) {
+                        try {
+                            $tenant = app()->call("$item@{$method->getName()}");
+                        } catch (BindingResolutionException $e) {
+                            // Prevent trying to find a tenant when bindings aren't working for them.
+                        }
+
+                        if ($tenant) {
```
Comment (ArlonAntonius) [inline src/Tenancy/Identification/TenantResolver.php:163]: Uhmmm this might actually be something that we want to add no matter what, if this PR is not going through, please make this a separate PR.

**Review by ArlonAntonius — COMMENTED** [review]
Communicated most of this to you on Discord, want to point it out over here so we know we have this communication documented.

File: src/Tenancy/Identification/TenantResolver.php (hunk 4 above)
Comment (luceos) [reply src/Tenancy/Identification/TenantResolver.php:163]: This code is very specific to this functionality, will separate it if this PR isn't merged.

File: src/Identification/Queue/Events/Processing.php (hunk 1 above)
Comment (luceos) [reply src/Identification/Queue/Events/Processing.php]: Agreed, refactored all of that last night.

**Comment by ArlonAntonius** [comment]
I suddenly realized an issue we had with multi-tenant. dispatch_now is not identifying tenants properly as the dispatcher doesn't go through the queue.

I'm not sure whether this should be considered for the queue identification or that this should be something on its own. This is something that needs to be addressed before the release though.

**Comment by morloderex** [comment]
@ArlonAntonius Well, if the consumer is dispatching an event via the dispatch_now function then they could just dispatch the new event from within another tenant aware queued job? :)

If someone wanted to use dispatch_now outside of an already queued event then i don't think we have a way for automatically detecting a tenant, we should ask the consumer to use the sync driver before doing any queuing at all and then just have them dispatch as normal without the even using the dispatch_now function.

Does that even make sense to you? :)

**Comment by luceos** [comment]
🎉
